src/GovUk.Education.ExploreEducationStatistics.Data.Services/TimePeriodService.cs

```diff
@@ -75,30 +66,35 @@ public TimePeriodLabels GetTimePeriodLabels(Guid subjectId)
                 TimePeriodLabelFormatter.Format(last.Year, last.TimeIdentifier));
         }
 
-        private static IList<(int Year, TimeIdentifier TimeIdentifier)> GetDistinctObservationTimePeriods(
-            IQueryable<Observation> observations)
+        private static async Task<IList<(int Year, TimeIdentifier TimeIdentifier)>> GetDistinctObservationTimePeriods(
+            IQueryable<Observation> observationsQuery)
         {
-            return observations
-                .Select(o => new {o.Year, o.TimeIdentifier})
+            var timePeriods = await observationsQuery
+                .Select(o => new Tuple<int, TimeIdentifier>(o.Year, o.TimeIdentifier))
                 .Distinct()
-                .AsNoTracking()
-                .ToList()
-                .OrderBy(tuple => tuple.Year)
-                .ThenBy(tuple => tuple.TimeIdentifier)
-                .Select(tuple => (tuple.Year, tuple.TimeIdentifier))
-                .ToList();
+                .ToListAsync();
+
+            return OrderTimePeriods(timePeriods);
         }
-        
+
         private static IList<(int Year, TimeIdentifier TimeIdentifier)> GetDistinctObservationTimePeriods(
             IList<Observation> observations)
         {
-            return observations
-                .Select(o => new {o.Year, o.TimeIdentifier})
+            var timePeriods = observations
+                .Select(o => new Tuple<int, TimeIdentifier>(o.Year, o.TimeIdentifier))
```

Comment:
Without an explicit `Tuple` type, we can define it like:

```cs
.Select(o => (o.Year, o.TimeIdentifier))
```

Same above as well

Reply:
I can't do that above because it's the `IQueryable` variant and LINQ has no expression-tree support for `ValueTuple`.

In the expression-tree above we can only use `Tuple` - `.Select(o => new Tuple<int, TimeIdentifier>(o.Year, o.TimeIdentifier))` or an anonymous type - `.Select(o => new { o.Year, o.TimeIdentifier })`.

If we went with an anonymous type we can't pass it as an argument and have a common `OrderTimePeriods` method.

That's why I ended up using `Tuple` for both variants and then wrote `OrderTimePeriods` accepting `IList<Tuple<int, TimeIdentifier>> timePeriods`. However as you've seen, it has a nasty conversion from `Tuple` to `ValueTuple` using `.Select(tuple => ((int Year, TimeIdentifier TimeIdentifier)) tuple.ToValueTuple())` to satisfy the return type and name the arguments.

I've got an alternative solution to incorporate your suggestions and in the `IQueryable` variant project the database result from anonymous type to `ValueTuple`:

```
  var timePeriods = (await observationsQuery
          .Select(o => new { o.Year, o.TimeIdentifier })
          .Distinct()
          .ToListAsync())
      .Select(x => (x.Year, x.TimeIdentifier))
      .ToList();
```

I'll add these changes as a separate commit and if you approve then I'll squash them. Thanks

Reply:
Of course, LINQ expressions strike again :nauseated_face: 

I like the new approach - nice to avoid the conversion!